src/jmteb/evaluators/clustering/evaluator.py

```diff
@@ -22,80 +24,95 @@
             )
-        if self.hierarchical_clustering:
-            clustering_models.extend(
-                [
-                    AgglomerativeClustering(n_clusters=n_clusters),
-                    BisectingKMeans(n_clusters=n_clusters),
-                    Birch(n_clusters=n_clusters),
-                ]
+            test_labels = [item.label for item in self.test_dataset]
+
+        n_clusters = len(set(test_labels))
+        clustering_models = {
+            type(model).__name__: model
+            for model in (
+                MiniBatchKMeans(n_clusters=n_clusters, n_init="auto"),
+                AgglomerativeClustering(n_clusters=n_clusters),
+                BisectingKMeans(n_clusters=n_clusters),
+                Birch(n_clusters=n_clusters),
             )
-        for clustering_model in clustering_models:
-            _results: dict[str, float] = {}
-            clustering_model.fit(embeddings)
-            y_pred = clustering_model.labels_
+        }
 
-            # compute metric
-            h_score, c_score, v_score = homogeneity_completeness_v_measure(
-                labels_pred=y_pred, labels_true=np.array(labels)
+        logger.info("Fitting clustering model...")
+        val_results = {}
+        for clustering_model_name, clustering_model in clustering_models.items():
+            val_results[clustering_model_name] = self._evaluate_clustering_model(
+                val_embeddings, val_labels, clustering_model
             )
-            clustering_model_name = type(clustering_model).__name__
-            _results.update(
-                {
-                    "v_measure_score": v_score,
-                    "homogeneity_score": h_score,
-                    "completeness_score": c_score,
-                }
+        optimal_clustering_model_name = sorted(
+            val_results.items(),
+            key=lambda res: res[1][self.main_metric],
+            reverse=True,
+        )[0][0]
+
+        test_results = {
+            optimal_clustering_model_name: self._evaluate_clustering_model(
+                test_embeddings,
+                test_labels,
+                clustering_models[optimal_clustering_model_name],
             )
-            results[clustering_model_name] = _results
+        }
 
         return EvaluationResults(
             metric_name=self.main_metric,
-            metric_value=max([v[self.main_metric] for v in results.values()]),
-            details=results,
+            metric_value=test_results[optimal_clustering_model_name][self.main_metric],
+            details={
+                "optimal_clustering_model_name": optimal_clustering_model_name,
+                "val_scores": val_results,
+                "test_scores": test_results,
+            },
+        )
+
+    @staticmethod
+    def _evaluate_clustering_model(embeddings: np.ndarray, y_true: list, clustering_model: Any) -> dict[str, float]:
+        clustering_model_ = deepcopy(clustering_model)
```

Comment:
```suggestion
    def _evaluate_clustering_model(embeddings: np.ndarray, y_true: list, model_constructor: Callable[[], ClusteringModel]) -> dict[str, float]:
        clustering_model_ = model_constructor()
```

clustering_model の Typing はがんばって次のようにしても良いと思いました。
https://stackoverflow.com/questions/54868698/what-type-is-a-sklearn-model